### PR TITLE
Fix: other courses

### LINKED ISSUE #44: Announce course

### LINKED ISSUE #48: Fix "Other courses" typo
Image

### PR DESCRIPTION
**Pull request by Illia-Sakharau**
## What type of PR is this? (select all that apply)

- [x] 🐛 Bug Fix

## Description
- [x] Return nearest course to Upcoming courses
- [x] Refactor component `Courses`
- [x] Update start date for courses
- [x] Change label text on course page (in `CourseMain` component)

## Related Tickets & Documents
- Related Issue #44 
- Closes #44 

- Related Issue #48 
- Closes #48 

## Screenshots, Recordings
Screenshot 2024-04-16 at 16 54 55

## Added/updated tests?

- [x] 👌 Yes
- [ ] 🙅‍♂️ No, because they aren't needed
- [ ] 🙋‍♂️ No, because I need help

### CODE REVIEW

**Review by andron13 — CHANGES_REQUESTED** [review]
1. The component seems quite large and does quite a few tasks. You might want to consider breaking it down into smaller components, for instance, CourseList and CourseItem.
2. I have to say, I found **courses as Course[]** to be less than ideal.
3. What's going to happen if the data array turns out to be empty?
4. We seem to be parsing the date twice over here. Is there a way we could avoid this duplicate use of `Date.parse`?

**Comment by Illia-Sakharau** [comment]
1. Removed all unnecessary items from the component.
2. Replaced with a data type check.
3. Added the processing of this case (empty data array).
4. This is not a duplicate of `Date.parse`, I will convert the two values being compared.

File: src/app/hooks/use-nearest-course/utils/choose-nearest-course.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  prevCourse?: Course,
+  nextCourse?: Course,
+  bufferPeriod: Day,
+}
+
+export function chooseNearestCourse({prevCourse, nextCourse, bufferPeriod}: chooseNearestCourseProps): Course | undefined {
+  const dateNow = Date.now();
+  let course = !nextCourse ? prevCourse : nextCourse;
+  if (nextCourse && prevCourse) {
+    const bufferPeriodMs = bufferPeriod * 24 * 60 * 60 * 1000;
```
Comment (andron13) [inline src/app/hooks/use-nearest-course/utils/choose-nearest-course.ts:14]: А зачем переменные создавать в проверке условия?

File: src/app/hooks/use-nearest-course/utils/define-prev-next.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      ) {
+        prevCourse = obj;
+      }
+      if (
+        !nextCourse && !isPast
+        || (!isPast && Date.parse((nextCourse as Course).startDate) > startDate)
+      ) {
+        nextCourse = obj;
+      }
+    }
```
Comment (andron13) [inline src/app/hooks/use-nearest-course/utils/define-prev-next.ts:28]: я не смог это прочитать.

File: src/features/courses/courses/courses.tsx [hunk 3]
```diff
@@ -1,55 +1,43 @@
...
         <div className="rs-courses-wrapper">
           {sortedCourses.map(
             (course) => {
-              if (course.id === nearestCourse?.id) return;
               return (
                 <CourseCard
                   key={course.id}
                   {...course}
-                  startDate={ dateNow < Date.parse(course.startDate) ? course.startDate : '(Upcoming)' }
+                  startDate={ nearestCourseStartDate <= Date.parse(course.startDate) ? course.startDate : '(TBD)' }
```
Comment (andron13) [inline src/features/courses/courses/courses.tsx:34]: я не приму в рендере расчёты.

File: src/app/hooks/use-nearest-course/utils/choose-nearest-course.ts (hunk 1 above)
Comment (Illia-Sakharau) [reply src/app/hooks/use-nearest-course/utils/choose-nearest-course.ts]: Изменил

File: src/features/courses/courses/courses.tsx (hunk 3 above)
Comment (Illia-Sakharau) [reply src/features/courses/courses/courses.tsx]: убрал расчеты из рендера

File: src/app/hooks/use-nearest-course/utils/define-prev-next.ts (hunk 2 above)
Comment (Illia-Sakharau) [reply src/app/hooks/use-nearest-course/utils/define-prev-next.ts:28]: Рефакторинг запланирован в таске [#85](https://github.com/rolling-scopes/site/issues/85)

**Review by dzmitry-varabei — APPROVED** [review]

**Review by natanchik — APPROVED** [review]

**Pull request merged**